perlclib: Add extensive advice for interfacing with libc

khwilliamson opened the pull request:
We should decide where this all should go if not perlxs.

tonycoz commented:
None of this is "XS language".

I think it belongs in perlguts or perlhacktips, or possibly a new pod.

The existing text under CAVEATS should probably move to the same place.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+the C<ENV_LOCK> for this purpose, provided that using it doesn't unduly
+slow down actual environment accesses.
+
+=item C<Ꞧ>
+
```
This is horribly mangled by the github pod viewer, and similarly in the table.

https://github.com/khwilliamson/perl5/blob/perlxs/dist/ExtUtils-ParseXS/lib/perlxs.pod

I don't know if adding an `=encoding` directive would help.  (perlpod says "...if your encoding isn't US-ASCII, then put a C<=encoding I<encodingname>> command very early in the document"

tonycoz replied:
Also, I expect it won't display properly by default in a terminal on Windows.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+call to C<tzset>, which writes to time-related global variables.
+Therefore they require an exclusive lock.  See C</L</M>>.
+
+=item C<×>
```
The `×` shows here fine in the github viewer, but is mangled in the table.

mauke commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+furnishes the buffer, and which is thread-safe.  Other flags on the
+entry should indicate this.
+
+=item E
```
Missing `C< >`.

mauke commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+
+line.  If you don't include F<perl.h>, you must instead place this line
+
+ #include <regen/reentr.pl>
```
Shouldn't this be `<reentr.h>`?

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+then the program is safe with regards to these.  If this is done, then
+read-only accesses can be protected by instead using
+
+ ENV_READ_LOCK
+ ...
+ ENV_READ_UNLOCK
```
It might be worth mentioning here that the return value of getenv() won't be valid after the `ENV_READ_UNLOCK;`

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+reading or writing the environment at the same time.  If all writers (as
+the core does) protect their accesses by surrounding them with
+
+ ENV_LOCK
```
All the `*_LOCK`/`*_UNLOCK` lines should have a terminating `;` since that's how they will appear in code.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+
+These time-related functions may do writes; perhaps through an indirect
+call to C<tzset>, which writes to time-related global variables.
+Therefore they require an exclusive lock.  See C</L</M>>.
```
Is the first `/` in `C</L</M>>` intended here?  It displays like:
![image](https://github.com/Perl/perl5/assets/166429/19893c11-9bc9-4eee-a263-cc9b7fc7965e)

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2112,27 +2114,410 @@ than a dMY_CTX in another source file.
+ dbm_store()
+ dirname()          B
+ dlerror()          B
+ drand48()          Ꞧ
```
Perl provides Perl_drand48() (not just on POSIX-ish)

Unfortunately it provides none of the other *rand48() functions.

tonycoz commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2427,3 +2814,5 @@ details so they are only useful to the maintainer of the XS code itself.
 Originally written by Dean Roehrich <F<roehrich@cray.com>>.
 
 Maintained since 1996 by The Perl Porters <F<perl5-porters@perl.org>>.
+=item C<×>
+ getlogin()         × å
```
I expect this should be removed.

khwilliamson commented:
On 10/4/23 17:10, Tony Cook wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In dist/ExtUtils-ParseXS/lib/perlxs.pod 
> <https://github.com/Perl/perl5/pull/21529#discussion_r1346569440>:
> 
>> +=item C<Ꞧ>
> +
> 
> This is horribly mangled by the github pod viewer, and similarly in the 
> table.
> 
> https://github.com/khwilliamson/perl5/blob/perlxs/dist/ExtUtils-ParseXS/lib/perlxs.pod <https://github.com/khwilliamson/perl5/blob/perlxs/dist/ExtUtils-ParseXS/lib/perlxs.pod>

The first line already is an =encoding line.  Things appear to display 
fine to me except in the verbatim lines, which must be some bug that I'm 
surprised hasn't been found before,.


> 
> I don't know if adding an |=encoding| directive would help. (perlpod 
> says "...if your encoding isn't US-ASCII, then put a C<=encoding I> 
> command very early in the document"
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/Perl/perl5/pull/21529#pullrequestreview-1658623549>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAA2DH63QDTSCYOGPNI22C3X5XUGDAVCNFSM6AAAAAA5IMTUR2VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTMNJYGYZDGNJUHE>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

tonycoz commented:
I was looking at https://github.com/khwilliamson/perl5/blob/perlxs/dist/ExtUtils-ParseXS/lib/perlxs.pod

![image](https://github.com/Perl/perl5/assets/166429/f19260ec-bd7f-46f9-b8cc-f0083427d151)
![image](https://github.com/Perl/perl5/assets/166429/6a63b68b-139f-4f63-b1c9-294cddcba716)

jkeenan commented:
@khwilliamson, do you wish us to proceed on this p.r.?  Thanks.

khwilliamson commented:
after it works

jkeenan commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2208,11 +2184,949 @@ In Makefile.PL add -ltirpc and -I/usr/include/tirpc.
+platform doesn't have this issue.  Otherwise, you may have to change
+more than one category to accomplish your task.
+
+The exception to you being called with the numeric parts of the locale
```
"The exception to you being called ..."  Wording is awkward.

jkeenan commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2208,11 +2184,949 @@ In Makefile.PL add -ltirpc and -I/usr/include/tirpc.
+
+If your code doesn't do any of these things, nor depends on any of their
+values, then Congratulations!!, you don't have to worry about MULTIPLICITY
+nor threading.  But wait, a surprising number of libc functions do
```
Stylistic nits:  I expect the word `nor` to be paired with `neither`.  I would word this sentence:
```
If your code doesn't do any of these things or depend upon any of their
values, then Congratulations!!, you don't have to worry about MULTIPLICITY
or threading.
```

tonycoz replied:
"nor" without "neither" is rare, but it is valid. See method 2 at https://www.wikihow.com/Use-Nor

If it were to be rephrased to avoid "nor" I think it would need to be:
```
If your code doesn't do any of these things and does not depend on any of their
values, then Congratulations!!, you don't have to worry about MULTIPLICITY
nor threading.
```

jkeenan replied:
Well, that still retains a 'nor'.  Shouldn't the last phrase be:
```
you don't have to worry about MULTIPLICITY
or threading.
```

khwilliamson replied:
I still like the original.  Being a dinosaur, I consulted Strunk and White "The Elements of Style" and Fowler's "Modern English Usage", and neither is not required for nor.  I prefer nor here because it emphasizes the negative aspect.

khwilliamson replied:
changed

jkeenan commented on dist/ExtUtils-ParseXS/lib/perlxs.pod:
```diff
@@ -2208,11 +2184,949 @@ In Makefile.PL add -ltirpc and -I/usr/include/tirpc.
+=head3 Functions that are thread-safe when called with appropriate arguments
+
+Some of the functions are thread-safe if called with arguments that
+comply with certain (easily-met) restrictions.  These are:
```
We don't need the hyphen there, just a wordspace.

jkeenan reviewed (dismissed):
"The exception to you being called ..."  Wording is awkward.

(In general I can make only small stylistic comments.  The C and POSIX discussed in these docs is something I've never learned.)

khwilliamson commented:
I think this should be a new pod, because it can be generated (and most of the lists were generated)  But what to name it?

tonycoz commented:
I could the thread parts going into perlthrtut (which has a small section on reentrancy) or perlguts (or a new perlthrguts)

The locale parts could go into perlguts or a new perllocguts.

khwilliamson commented:
Much of the data for this PR came from a __DATA__ section for generating another file for 5.41.  I think it would be best to make the contents of this PR be able to be generated from that same DATA section, to avoid redundancy, and make it easier to update as needed.  I'm ok with separating threads vs non-threads, but I think the locale stuff shouldn't be separated from the non-locale stuff, because it is too easy to think, "I'm not using locales" and skip reading a separate file, whereas they may very well apply to you.  Note that the text of this applies to core development as well as XS modules.

tonycoz commented:
> Note that the text of this applies to core development as well as XS modules.

It could all go in perlguts, which is generally things relevant to both XS and core.

khwilliamson committed:
Add extensive guidance to perlclib

This consolidates much of the pod about interfacing with the standard C
library into this pod, while adding extensive documentation.

khwilliamson commented:
I started to put this in perlguts, but the description of that file doesn't really fit this: 

> This document attempts to describe how to use the Perl API, as well as
to provide some info on the basic workings of the Perl core.

I had forgotten about `perlclib`, which some of this PR duplicates, but perlclib had more extensive advice.  So, I've modified the PR to instead add things to that pod.
